### PR TITLE
check more case for empty binding patten

### LINKED ISSUE #24634: Destructuring assignment doesn't check assignability
A value being destructed by `[]` must be iterable.
A value being destructed by `{}` must be non-nullable.

**TypeScript Version:**  master

**Search Terms:** 

**Code**

```ts
const [] = {}; // should be error
const {} = undefined; // error correctly
(([]) => 0)({}); // should be error
(({}) => 0)(undefined); // should be error
```

**Expected behavior:**

**Actual behavior:**

**Playground Link:**  

**Related Issues:**

### PR DESCRIPTION
**Pull request by Kingwl**
Fixes #24634

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -23664,16 +23664,30 @@ namespace ts {
             }
             // For a binding pattern, validate the initializer and exit
             if (isBindingPattern(node.name)) {
-                // Don't validate for-in initializer as it is already an error
-                if (node.initializer && node.parent.parent.kind !== SyntaxKind.ForInStatement) {
-                    const initializerType = checkExpressionCached(node.initializer);
-                    if (strictNullChecks && node.name.elements.length === 0) {
-                        checkNonNullType(initializerType, node);
+                const needCheckInitializer = node.initializer && node.parent.parent.kind !== SyntaxKind.ForInStatement;
+                const needCheckWinendType = node.name.elements.length === 0;
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Spelling

**Comment by RyanCavanaugh** [comment]
@Kingwl ping on this one - would be ready to merge after fixes

**Comment by Kingwl** [comment]
🙋🏻‍♂️

**Comment by RyanCavanaugh** [comment]
@Kingwl I promise to merge after the conflict is resolved! 👋

**Comment by Kingwl** [comment]
@typescript-bot test this.

**Pull request merged**